File: a_star.cpp
#include<iostream>
#include<vector>
#include<algorithm>
#include <cfloat>
#include<set>

using namespace std;


/*
Let's learn about A star Algorithm for path finding:

We have a grid of MxN which consists of 0s and 1s
0s represent a free path, whereas 1s represent obstacle, so the path cannot pass through that cell.alignas

We have coordinates of the start node and the target node.

Our job is to used informed search technique to reach the target node, without exploring all the options (without being Greedy!)

For that we will use a heuristic cost function "h" that will give us the approximate/estimated cost from current Node to the target Node.alignas
Here, we have choice to use any function, but for simplicity and efficiency, let's use the Manhattan distance.

MANHATTAN DISTANCE:

manhattan(currentNode, targetNode) -> absolute_difference(currentNode.x, targetNode.x) + absolute_difference(currentNode.y, targetNode.y)

Along with this, we need to combine the actual cost "g"
Together with g and h, we get the total cost "f"

Starting with 0 at startNode, we just need to increment g by 1 after each move.

EXPLORING THE NEAGHBORS:

Here, let's allow movement only row wise and column wise. Therefore we can change the position by 4 combinations in total:
1. (1, 0)
2. (-1, 0)
3. (0, 1)
4. (0, -1)

Going further, if we want to allow diagonal movements, we can also include (1, 1), (1, -1), (-1, 1), and (-1, -1). 

EXAMPLE:

Grid (0=free, 1=obstacle):
0 1 0 0 0 
0 1 0 1 0 
0 1 0 1 0 
0 0 0 1 0 
0 0 0 0 0 

Type 1: Only vertical and horizontal movement allowed: 

Finding path from (0,0) to (4,4)
(0,0) (1,0) (2,0) (3,0) (4,0) (4,1) (4,2) (4,3) (4,4)

Type 2: Diagonal movement allowed:

Finding path from (0,0) to (4,4)
(0,0) (1,0) (2,0) (3,1) (4,2) (4,3) (4,4) 


*/


class Node{
    public:
    int x, y;
    float h, g, f;
    Node* parent;

    Node(int x, int y, float h = FLT_MAX, float g = FLT_MAX){
        this->x = x;
        this->y = y;
        this->h = h;
        this->g = g;
        this->f = g + h;
        this->parent = nullptr;
    }
};

const int dx[] = {1, -1, 0, 0, 1, 1, -1, -1};
const int dy[] = {0, 0, 1, -1, 1, -1, 1, -1};

int heuristic(Node *a, Node *b){
    return abs(a->x - b->x) + abs(a->y - b->y);
}

bool isValid(int x, int y, vector<vector<int> >&grid){
    return (x >= 0 && y >= 0 && x < grid.size() && y < grid[0].size() && grid[x][y] == 0);
}

struct CompareNodes {
    bool operator()(Node* a, Node* b) {
        return a->f > b->f;  // min-heap behavior
    }
};

bool reachedGoal(Node *a, Node *target){
    return (a->x == target->x && a->y == target->y);
}

void aStar(vector<vector<int> >&grid, pair<int, int> &start, pair<int, int> &target){

    priority_queue<Node*, vector<Node*>, CompareNodes> openList;    

    set<pair<int, int>> closedList;

    Node* startNode = new Node(start.first, start.second);
    Node* targetNode = new Node(target.first, target.second);
    startNode->g = 0;
    startNode->h = heuristic(startNode, targetNode);
    openList.push(startNode);

    while(!openList.empty()){

        Node* current = openList.top();
        openList.pop();


        if(reachedGoal(current, targetNode)){
            vector<pair<int, int>> path;
            while(current != nullptr){
                path.push_back({current->x, current->y});
                current = current -> parent;
            }
            reverse(path.begin(), path.end());
            for(auto p : path){
                cout<<"("<<p.first<<","<<p.second<<") ";
            }
            cout<<endl;
            return;
        }

        
        closedList.insert({current->x, current->y});

        for(int i=0; i<8; i++){
            int newx = current->x + dx[i];
            int newy = current->y + dy[i];

            if(isValid(newx, newy, grid) && closedList.find({newx, newy})==closedList.end()){
                Node* neighbor = new Node(newx, newy);
                neighbor->g = current->g + 1;
                neighbor->h = heuristic(neighbor, current);
                neighbor->f = neighbor->g + neighbor->h;
                neighbor->parent = current;

                openList.push(neighbor);
            }
        }
    }

    cout<<"NO PATH FOUND!"<<endl;
}

int main() {
    int rows = 5, cols = 5;
    vector<vector<int>> grid(rows, vector<int>(cols, 0));
    
    grid[0][1] = 1;
    grid[1][1] = 1;
    grid[2][1] = 1;
    grid[1][3] = 1;
    grid[2][3] = 1;
    grid[3][3] = 1;
    
    cout << "Grid (0=free, 1=obstacle):\n";
    for(int i = 0; i < rows; i++) {
        for(int j = 0; j < cols; j++) {
            cout << grid[i][j] << " ";
        }
        cout << endl;
    }
    
    pair<int, int> start = {0, 0}; 
    pair<int, int> target = {4, 4};
    
    cout << "\nFinding path from (" << start.first << "," << start.second << ") to ("
         << target.first << "," << target.second << ")\n\n";

    aStar(grid, start, target);
    
    return 0;
}

File: a_star.java
import java.util.*;

// Thanks @tirthraj07 - I am now convinced to give up on STL :)

class Node implements Comparable<Node>{
    // A Node will have x and y coordinates, values of g, h and f, and the parent Node

    public int x, y;
    public double g, h, f;
    public Node parent;

    // let's define the constructor

    public Node(int x, int y){
        this.x = x;
        this.y = y;
        this.g = Double.MAX_VALUE;
        this.h = Double.MAX_VALUE;
        this.f = Double.MAX_VALUE;
        this.parent = null;

        // now, we need to override the methods that are defined in the interface of Comapable<Node>  
    }

    // check of the x and y of this Node and the other node are equal. Before that, we firstly check if the passed Object is an instace of class Node


    @Override
    public boolean equals(Object obj){
        if(obj instanceof Node other){
            return this.x == other.x && this.y == other.y;
        }
        return false;
    }

    // in compareTo, we check compare the f values of the current and other Node

    @Override
    public int compareTo(Node other){
        return Double.compare(this.f, other.f);
    }

    // the default hashCode method will generate unique hashcodes for each Object. But here, we need to override it to generate same hashcodes for Nodes with same x and y

    @Override
    public int hashCode(){
        return Objects.hash(this.x, this.y);
    }

    // we also override the toString method to print the object of Node

    @Override
    public String toString(){
        return "(" + this.x + ", " + this.y + ")";
    }
}

class AStar{

    // in this class, we need to set all the variables and data structures required for the A* algorithm

    private int NUM_ROW, NUM_COL;
    private int[] dx = {0, 0, 1, -1};
    private int[] dy = {1, -1, 0, 0};
    private int[][] grid;
    private PriorityQueue<Node> openList;
    private HashSet<Node> closedList;

    // we can also store the goalNode as a data member of AStar, as it wont change. We don't need to pass the goalNode all the time explicitly.

    private Node goal;

    // lets begin the actual implementation

    //we need to keep 3 methods at handy: heuristic function, path printing logic, validity if coordinates

    private int heuristic(Node other){
        return Math.abs(goal.x - other.x) + Math.abs(goal.y - other.y);
    }

    private boolean isValid(int x, int y){
        return x>=0 && y>=0 && x<NUM_ROW && y<NUM_COL && grid[x][y]!=1;
    }

    private void printPath(){
        List<Node> path = new ArrayList<>();
        Node curr = this.goal;
        while(curr != null){
            path.add(curr);
            curr = curr.parent;
        }
        Collections.reverse(path);
        for(Node item : path){
            System.out.print(item.toString() + " ");
        }
    }

    // we wil implement the logic in the constructor itself

    public AStar(int [][] grid, Node start, Node goal){

        

        this.grid = grid;
        this.NUM_ROW = grid.length;
        this.NUM_COL = grid[0].length;

        // added edge cases

        if(!isValid(start.x, start.y)){
            System.out.println("Invalid start coordinate!");
            return;
        }

        if(!isValid(goal.x, goal.y)){
            System.out.println("Invalid goal coordinate!");
            return;
        }
        
        this.goal = goal;
        this.openList = new PriorityQueue<>();
        this.closedList = new HashSet<>();



        start.g = 0;    // we start here, so actual distance from previous node is 0, as there is no previous node
        start.h = heuristic(start);     // calculate estimated cost from start to goal (remember, goal is implicit here)
        
        start.f = start.h + start.g;    // set the total cost (h + g)

        openList.offer(start); 


        while(!openList.isEmpty()){

            Node current = openList.poll();
            closedList.add(current);

            // firstly, lets check if the goal is reached

            if(current.equals(goal)){
                goal.parent = current.parent;
                printPath();
                return;
            }

            // now we can explore the neighbors of current node, by moving is all the 4 directions

            for(int i=0; i<4; i++){
                int newx = current.x + dx[i];
                int newy = current.y + dy[i];

                //if newx, newy are out of bound or is a obstacle, skip it

                if(!isValid(newx, newy)) continue;

                Node nbr = new Node(newx, newy);
                nbr.g = current.g + 1;  //g will be 1 more than the current node, as we move in unit distance
                nbr.h = heuristic(nbr);
                nbr.f = nbr.g + nbr.h;
                nbr.parent = current;   //nbr is a child of current, so we set parent of nbr as current
                
                // now let's first check if it is the closedlist, if yes no need to revisit it

                if(closedList.contains(nbr)) continue;

                // now we also need to check that for same x, y do we have a Node in the openList already
                // if we have one, and it has less g value that nbr, we remove it from the openList (THIS IS THE EXACT REASON TO NOT USE priority_queue in C++, as we cannot pop arbitrary item from it :/ )

                Node nodeAlreadyInOpenList = openList.stream().filter(n -> n.equals(nbr)).findFirst().orElse(null);

                if(nodeAlreadyInOpenList!=null && nodeAlreadyInOpenList.g > nbr.g){
                    openList.remove(nodeAlreadyInOpenList);
                }
                openList.offer(nbr);
            }
        }

        System.out.println("Goal is unreachable!");
    }

    public static void main(String[] args){
        int[][] grid = {
            {0, 0, 1, 1, 0},
            {1, 0, 0, 1, 1},
            {0, 1, 0, 0, 0},
            {1, 1, 0, 1, 1},
            {1, 0, 0, 0, 0}
        };

        Node startNode = new Node(0, 0);
        Node goalNode = new Node(4, 4);

        AStar aStar = new AStar(grid, startNode, goalNode);
    } 
}

